xdsl/frontend/passes/desymref.py

```diff
@@ -0,0 +1,389 @@
+# pseudocode, that would look like:
+#
+# ```
+# desymref_region(region):
+#   for block in region.blocks:
+#     for op in block.ops:
+#       desymref_op(op)
+#     # at this point no op contains a symref operation
+#     # in the nested region.
+# ```
+#
+# Next, we describe how desymrefication actually works, on a high level.
+# Every region either declares a symbol or uses it from the parent region.
+# Let's consider each case separately.
+#
+# Case 1: Symbol is declared in this region. This is an easy case. We already
+# know that symbol is only used in this region and no other nested region uses
+# it. Therefore running any SSA-construction algorithm on the CFG is enough.
+#
+# Case 2: Symbol is not declared in this region. In this case, we only support
+# non-CFGs at the moment (i.e. single block regions). We observe that in general
+# every symbol can be simplified within a block to have at most one fetch (start
+# of the block) at most one update (end of the block). This means that using
+# specification of an operation we can promote these symbols outside.
+# TODO: Add op promotion.
+
+
+def is_definition(op: Operation) -> bool:
+    return isinstance(op, symref.Declare)
+
+
+def is_use(op: Operation) -> bool:
+    return isinstance(op, symref.Fetch) or isinstance(op, symref.Update)
+
+
+def is_read(op: Operation) -> bool:
+    return isinstance(op, symref.Fetch)
+
+
+def is_write(op: Operation) -> bool:
+    return isinstance(op, symref.Update)
+
+
+def get_symbol(op: Operation) -> str:
+    """
+    Returns a symbol attribute for an operation. If operation has no symbol
+    attributes, raises an exception.
+    """
+    if isinstance(op, symref.Declare):
+        return op.sym_name.data
+    elif isinstance(op, symref.Fetch) or isinstance(op, symref.Update):
+        return op.symbol.root_reference.data
+    else:
+        raise FrontendProgramException(
+            f"Operation '{op.name}' does not have symbol attributes.")
+
+
+def get_symbols(block: Block) -> Set[str]:
+    """Returns a set of all symbols defined/used in a basic block."""
+    symbols: Set[str] = set()
+    for op in block.ops:
+        if is_definition(op) or is_use(op):
+            symbols.add(get_symbol(op))
+    return symbols
+
+
+def count_ops_by(block: Block, cond: Callable[[Operation], bool]) -> int:
+    """
+    Returns the number of operations in the block satisfying the given
+    condition.
+    """
+    count = 0
+    for op in block.ops:
+        if cond(op):
+            count += 1
+    return count
+
+
+def select_ops_by(block: Block, cond: Callable[[Operation],
+                                               bool]) -> List[Operation]:
+    """
+    Returns a list of operations in the block satisfying the given condition.
+    """
+    selected: List[Operation] = []
+    for op in block.ops:
+        if cond(op):
+            selected.append(op)
+    return selected
+
+
+def lower_bound(ops: List[Operation], op: Operation) -> Operation | None:
```

Comment:
This lower_bound is only defined in terms of operation order in the IR, right? To me, this felt a bit weird, I initially thought we were in some kind of lattice of operation types or whatever.

Maybe rename it to `lower_positional_bound` or something...